### PR TITLE
Implement opt-in change detection bypass

### PR DESCRIPTION
**Pull request by TheRawMeatball**
# Objective
Implements #2348  

## Solution
Implements the solution outlined [here](https://github.com/bevyengine/bevy/issues/2348#issuecomment-864445217)

### CODE REVIEW

File: crates/bevy_ecs/src/change_detection.rs [hunk 1]
```diff
@@ -41,6 +41,9 @@ pub trait DetectChanges {
     fn set_changed(&mut self);
 }
 
+// TODO: Add some nice docs here
```
Comment (TheRawMeatball) [inline crates/bevy_ecs/src/change_detection.rs:44]: If someone would contribute to this point, it'd be highly appreciated

**Comment by Ratysz** [comment]
This looks delightfully overengineered. I would be on board, but my sentiment from linked issue remains: do we actually need it?

The two examples so far can be (and best be, if you ask me) solved by splitting resources/components not just by the data they store, but also by the mutation events that happen to them, if that makes sense. I think that's a universal solution, be it inside a specific plugin or between several plugins.

**Comment by NathanSWard** [comment]
> but also by the mutation events that happen to them

@Ratysz Would you mind clarifying exactly what you mean by this?

**Comment by Ratysz** [comment]
> > but also by the mutation events that happen to them
> 
> @Ratysz Would you mind clarifying exactly what you mean by this?
> (I assume it just mean on when they trigger change detection via `DerefMut` for example?)

Yes, pretty much. E.g., we could split `Assets` into `AssetStorage` and `AssetEvents`, separating the concern of providing assets from that of processing asset-related events.

**Comment by TheRawMeatball** [comment]
While this can sometimes be an option, I don't think it can be applied globally as it requires changing the external API to do something that only bothers the internal impl. For the asset case for example, the external API shouldn't need to care about how the `Assets` type also accumulates events and sends them back.  This impl allows for building such abstractions.

**Comment by alice-i-cecile** [comment]
Ran into [a user with a serious need for this earlier today](https://discordapp.com/channels/691052431525675048/742569353878437978/917848513801519184). Effectively, they wanted to make a change detecting that system that responded to all changes other than its own. I spent a while helping them and couldn't come up with a clean workaround.

This now has my 👍🏽 in terms of use case. Lets take a look at the specific design now...

File: crates/bevy_ecs/src/change_detection.rs (hunk 1 above)
Comment (alice-i-cecile) [reply crates/bevy_ecs/src/change_detection.rs:44]:
```rust
/// When implemented for a component or resource type, this trait allows end users to manually bypass change detection within the ECS using the [`get_untracked`] method on `Query` and `World`.
```

I *think* that's where those methods live?

**Comment by alice-i-cecile** [comment]
Gave you some docs ;) The impl looks reasonable to me. Two outstanding concerns:

1. From an API perspective, I would also be interested in supporting `Query<Untracked<& mut A>>`, which better communicates the intent in the function signature and should allow ordinary iteration in ways that the current design does not.
2. This is reasonably non-trivial: it should have tests to verify that it acts as promised.

File: crates/bevy_ecs/src/change_detection.rs (hunk 1 above)
Comment (MrGVSV) [reply crates/bevy_ecs/src/change_detection.rs:44]: I don't see the `get_untracked` methods in [`Query`](https://github.com/bevyengine/bevy/blob/cf48132efd107bbca85aaea00589a01457ef4872/crates/bevy_ecs/src/system/query.rs) or [`World`](https://github.com/bevyengine/bevy/blob/cf48132efd107bbca85aaea00589a01457ef4872/crates/bevy_ecs/src/world/mod.rs).

Are they found elsewhere? Or are they part of an unmerged PR?

**Comment by MrGVSV** [comment]
I haven't used `Reflect` all that much, but is there a reason it doesn't have the other impl macros? (i.e., `impl_into_inner`, `impl_change_detection_circumvention`, `impl_debug`)
https://github.com/bevyengine/bevy/blob/6591742a73e9632bdaebe9d8ac9cda969b0e4ef3/crates/bevy_ecs/src/change_detection.rs#L216

File: crates/bevy_ecs/src/change_detection.rs (hunk 1 above)
Comment (alice-i-cecile) [reply crates/bevy_ecs/src/change_detection.rs:44]: It was a "please add these methods too" comment for @TheRawMeatball ;)

**Comment by TheRawMeatball** [comment]
Closing in favor of the upcoming permissions work

**Comment by alice-i-cecile** [comment]
I ran into a very compelling use case of this today.

I have two components: `Direction` and `Rotation`, that I want to keep synchronized. If one has changed, update the other to match.

However, this creates an infinite ping-pong, as doing so causes the other to be changed, thus triggering another update. The workaround is to check whether the new value is equal to the old value before setting it.

**Comment by alice-i-cecile** [comment]
Reopening this: permissions has serious open questions, and this is at least an important stop-gap.

**Pull request closed without merging**